### PR TITLE
ML: Add support for single bucket aggs in Datafeeds

### LINKED ISSUE #36838: [ML] Support single bucket aggregations in datafeeds
This use case was described in the discuss forum [Referencing field name from datafeed aggregation to use as a detector in an ML job 
](https://discuss.elastic.co/t/referencing-field-name-from-datafeed-aggregation-to-use-as-a-detector-in-an-ml-job/160638) where it is desired to use a filter aggregation in the datafeed.

```
  "aggs": {
    "metric1":{
      "filter":{"term":{"my_filter_field":"A"}},
      "aggs":{
        "metric1_value":{"value_count":{"field":"my_filter_field"}}
      }
    }
```

The canonical example in the docs is 
```
    "aggs" : {
        "t_shirts" : {
            "filter" : { "term": { "type": "t-shirt" } },
            "aggs" : {
                "avg_price" : { "avg" : { "field" : "price" } }
            }
        }
    }
```

The solution is to find single bucket value aggs in [AggregationToJsonProcessor](https://github.com/elastic/x-pack-elasticsearch/blob/master/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AggregationToJsonProcessor.java#L108) and handle them as if they were leaf node aggregations.

### PR DESCRIPTION
**Pull request by benwtrent**
Adds support for Single Bucket aggregations in DatafeedConfig 

Restrictions:

* We verify at each call of `processAggs` that the current level, and the next levels (as dictated by `SingleBucketAggregation` have a max of 1 `MultiBucketAggregation`
* We recursively handle each `SingleBucketAggregation` as the true leaf aggregation will be a sub aggregation. 

closes #36838

### CODE REVIEW

**Review by davidkyle — COMMENTED** [review]
Who would have thought it would all be so simple. Recursion FTW.

It would be nice to test some different cases is it possible to add some unit test coverage? `AggregationToJsonProcessorTests` uses a lot of mocking and I suspect the line `.filter(MultiBucketsAggregation.class::isInstance)` will fail with mocked objects so it may not be easy.

**Comment by davidkyle** [comment]
cc @richcollier This change enables filter aggregations in datafeeds see the linked issue and discuss thread for details. 

I wonder where we should document this, perhaps add it to the examples [here](https://www.elastic.co/guide/en/elastic-stack-overview/6.5/ml-configuring-aggregation.html)

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AggregationToJsonProcessor.java [hunk 1]
```diff
@@ -93,18 +94,30 @@ private void processAggs(long docCount, List<Aggregation> aggregations) throws I
...
                 bucketAggregations.add((MultiBucketsAggregation)agg);
+            } else if (agg instanceof SingleBucketAggregation){
+                singleBucketAggregations.add((SingleBucketAggregation)agg);
             } else {
                 leafAggregations.add(agg);
             }
         }
 
-        if (bucketAggregations.size() > 1) {
+        // If on the current level (indicated via bucketAggregations) or on of the next levels (singleBucketAggregations)
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AggregationToJsonProcessor.java]: typo: "or on of the next levels" needs correction

**Comment by benwtrent** [comment]
run docbldesx

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AggregationToJsonProcessor.java [hunk 2]
```diff
@@ -93,18 +95,39 @@ private void processAggs(long docCount, List<Aggregation> aggregations) throws I
...
         // The leaf aggregations will be processed first.
         for (Aggregation agg : aggregations) {
             if (agg instanceof MultiBucketsAggregation) {
                 bucketAggregations.add((MultiBucketsAggregation)agg);
+            } else if (agg instanceof SingleBucketAggregation){
+                // Skip a level down for single bucket aggs, if they have a sub-agg that is not
+                // a bucketed agg we should treat it like a leaf in this bucket
+                SingleBucketAggregation singleBucketAggregation = (SingleBucketAggregation)agg;
+                for (Aggregation subAgg : singleBucketAggregation.getAggregations()) {
+                    if (subAgg instanceof MultiBucketsAggregation || subAgg instanceof SingleBucketAggregation) {
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/AggregationToJsonProcessor.java]: I wonder if the test should be `subAgg instanceof HasAggregations`

**Review by davidkyle — APPROVED** [review]
LGTM. Thanks for adding the example to the docs

**Pull request merged**